Codechange: Replace all uses of SmallMap with std::map

PeterN opened the pull request:
## Motivation / Problem

SmallMap.

## Description

Replace all uses of SmallMap with std::map. Some other adjustments were needed to make this possible:

1) Replacing a pointer with a unique_ptr in FontColourMap.
2) Swapping key/value around for SocketList due to const constraints (SocketList was modifying its keys), 

## Limitations

Given the changes to SocketList, this probably shouldn't be a map in the first place?


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

PeterN committed:
Codechange: Use std::map instead of custom SmallMap.

PeterN committed:
Codechange: Swap SocketList map key/value around.

This map is used store socket and address together, and, other than
checking that the address does not already have a socket, the data layout
does not seem particularly important.

However, as address is the key, technically it should not be modified,
and address may self-modify itself during comparisons.

TrueBrain commented on src/fios.h:
```diff
@@ -16,6 +16,7 @@
 #include "gamelog.h"
 #include "network/core/tcp_content_type.h"
 #include "timer/timer_game_calendar.h"
+#include <map>
```
I think we should add this to `stdafx.h`. As the fact it works for some platforms and not for others is just annoying, and we shouldn't want to deal with :)

PeterN replied:
Seems reasonable. We did it for vector already :)

PeterN committed:
Fix: Don't rely on static initialization to set up sprite font caches.

The order of static initialization is undefined, so this can cause initalization
before relevant caches are initializations.

rubidium42 commented on src/fios.h:
```diff
@@ -26,7 +25,7 @@ enum SaveLoadInvalidateWindowData {
 };
 
-typedef SmallMap<uint, CompanyProperties *> CompanyPropertiesMap;
+using CompanyPropertiesMap = std::map<uint, CompanyProperties *>;
```
Rabbit hole-ish: `std::unique_ptr`?

PeterN replied:
Good point, I did it for the font map one already.

rubidium42 reviewed (commented):
I'd make the static initialisation and the flip of the socket list separate commits. The rest could be squashed into one commit I'd say.

rubidium42 commented on src/os/macosx/string_osx.cpp:
```diff
@@ -103,7 +103,7 @@ class CoreTextParagraphLayout : public ParagraphLayouter {
 				CFRange chars = CTRunGetStringRange(run);
 				auto map = fontMapping.begin();
-				while (map < fontMapping.end() - 1 && map->first <= chars.location) map++;
+				while (std::next(map) != fontMapping.end() && map->first <= chars.location) map++;
```
Wouldn't this be equal to `auto map = fontMapping.upper_bound(chars.location)`?

PeterN committed:
Codechange: Use unique_ptr for text layout font mapping.

This must stay a pointer as the value passed to other structures.

PeterN committed:
Codechange: Use unique_ptr for fios company properties.

This is the data used to show company names in the saveload window.
